Handle unknown fields with EXCLUDE, INCLUDE or RAISE

ramnes opened the pull request:
This is a rework of #595 so that it supports the API discussed in #524.

By default, only the fields described in the schema are returned when data is deserialized.

This commit implements an `unknown` option, so that `Schema(unknown=ALLOW)` or `Schema().load(data, unknown=ALLOW)` permit users to receive the unknown fields from the data.

`Schema(unknown=RAISE)` or `Schema().load(data, unknown=RAISE)` will raise a `ValidationError` for each unknown field.

`Schema(unknown=IGNORE)` or `Schema().load(data, unknown=IGNORE)` keep the original behavior.

Edit: `ALLOW` will finally be `INCLUDE`, and `IGNORE` will be `EXCLUDE`.

ramnes committed:
Handle unknown fields with INCLUDE, EXCLUDE, or RAISE

By default, only the fields described in the schema are returned when data is
deserialized.

This commit implements an `unknown` option, so that `Schema(unknown=INCLUDE)`
or `Schema().load(data, unknown=INCLUDE)` permit users to receive the unknown
fields from the data.

`Schema(unknown=RAISE)` or `Schema().load(data, unknown=RAISE)` will raise a
`ValidationError` for each unknown field.

`Schema(unknown=EXCLUDE)` or `Schema().load(data, unknown=EXCLUDE)` keep the
original behavior.

ramnes commented on marshmallow/marshalling.py:
```diff
@@ -275,6 +283,22 @@ def deserialize(self, data, fields_dict, many=False, partial=False,
+                        data=value,
+                        field_name=key,
+                        field_obj=type(value)
+                    )
```
I'm not very happy that I used `call_and_store` in a hackish way, but I could not find a better solution to achieve the same thing.

lafrech replied:
Maybe the logic that stores the error can be extracted in a `store_error` method and `call_and_store` can be changed call `store_error` upon `ValidationError`.

```python
    def call_and_store(self, getter_func, data, field_name, field_obj, index=None):
        try:
            return getter_func(data)
        except ValidationError as err:
            self.store_error(...)

    def store_error(...):
            self.error_kwargs.update(err.kwargs)
            self.error_fields.append(field_obj)
            self.error_field_names.append(field_name)
            errors = self.get_errors(index=index)
            # Warning: Mutation!
            if isinstance(err.messages, dict):
                errors[field_name] = err.messages
            elif isinstance(errors.get(field_name), dict):
                errors[field_name].setdefault(FIELD, []).extend(err.messages)
            else:
                errors.setdefault(field_name, []).extend(err.messages)
            # When a Nested field fails validation, the marshalled data is stored
            # on the ValidationError's valid_data attribute
            value = err.valid_data or missing
```

Then you can call `store_error` directly. And at this point, we could be tempted to remove `call_and_store` completely.

Also, I'm not sure about `field_obj=type(value)`. We might as well pass `None` or some arbitrary constant. And `store_error` could be modified to not append to `error_fields` if `None`, for instance.

Question is, what do we want to store in the `ValidationError` `fields` and `field_names`? Currently, we only have field errors or schema errors. Error for unknown keys are not really schema errors, but they don't correspond to a field either. Passing the data key as field name makes sense but I don't know what to pass as field object.

I don't think I ever used `field_names` and `fields` anyway. I've always been using `messages`. So I'm not sure what to expect from them.

If we decide not to append to `fields` and `field_names` for unknown key error, then user can't rely on their length being zero to assert there was no error. (But who would do that anyway?)

@sloria, @deckar01, thoughts?

sloria replied:
Hm, maybe we should factor out the error storage code into its own method, so this could look something like:

```python
self.store_error(
    'Unknown field.',
    data=value,
    field_name=key,
)
```

ramnes replied:
Thank you guys, I was a bit to shy to change this but that's indeed a better way of doing it.

ramnes replied:
Done. What do you both think of it that way?

lafrech commented on marshmallow/fields.py:
```diff
@@ -365,7 +366,8 @@ class Nested(Field):
         will be marshalled. If a field name (string) is given, only a single
         value will be returned as output instead of a dictionary.
         This parameter takes precedence over ``exclude``.
-    :param bool many: Whether the field is a collection of objects.
```
Did you mean to remove this line?

ramnes replied:
Nop, thanks! I think I've deleted this during the rebase of my old PR on the current HEAD of dev, which wasn't that easy. It's back now.

lafrech commented on marshmallow/utils.py:
```diff
@@ -18,6 +18,9 @@
 
 from marshmallow.compat import binary_type, text_type
 
+ALLOW = "allow"
```
It would be simpler to use plain strings, like `'allow'`, `'ignore'` and `'raise'`, rather than having to import those globals. I don't see the benefit of the globals.

ramnes replied:
@sloria what's your opinion on this? I was just following your proposition there, and I feel it's a great one, but honestly I'm good with both solutions.

ramnes replied:
Also, note to myself: if we keep it that way, I should replace double quotes by single quotes there.

sloria replied:
@lafrech Using constants prevents typos from passing silently, e.g. `MySchema(unknown='rais')`.

lafrech replied:
The point was to avoid having to import the constant. But I don't mind. In most use cases, this will be setup once in a base `Schema` class for the whole application.

sloria commented on tests/test_deserialization.py:
```diff
@@ -1315,6 +1315,39 @@ class MySchema(Schema):
+        with pytest.raises(ValidationError) as excinfo:
+            MySchema(unknown=RAISE).load({'foo': 3, 'bar': 5})
+        assert 'bar' in str(excinfo)
+
```
The code isn't handling list deserialization. The following test fails:

```python
        with pytest.raises(ValidationError) as excinfo:
            MySchema(unknown=RAISE, many=True).load([
                {'foo': 'abc'},
                {'foo': 3, 'bar': 5},
            ])
        err = excinfo.value
        assert 0 in err.messages
        assert 'foo' in err.messages[0]
        assert err.messages[0]['foo'] == ['Not a valid integer.']
        assert 1 in err.messages
        assert 'bar' in err.messages[1]
        assert err.messages[1]['bar'] == ['Unknown key.']
```

```
|| ______________________________________________________________________ TestSchemaDeserialization.test_unknown_fields_deserialization _______________________________________________________________________
|| tests/test_deserialization.py:1360: in test_unknown_fields_deserialization
||     assert 1 in err.messages
|| E   AssertionError: assert 1 in {0: {'foo': ['Not a valid integer.']}, 'bar': ['Unkown key.']}
|| E    +  where {0: {'foo': ['Not a valid integer.']}, 'bar': ['Unkown key.']} = ValidationError({0: {'foo': ['Not a valid integer.']}, 'bar': ['Unkown key.']},).messages
|| ========================================================================================= 1 failed in 0.11 seconds =========================================================================================
```

ramnes replied:
Fixed, thanks.

sloria commented on tests/test_deserialization.py:
```diff
@@ -1315,6 +1315,39 @@ class MySchema(Schema):
+
+        with pytest.raises(ValidationError) as excinfo:
+            MySchema(unknown=RAISE).load({'foo': 3, 'bar': 5})
+        assert 'bar' in str(excinfo)
```
This can be more specific:

```python
err = excinfo.value
assert 'bar' in err.messages
assert err.messages['bar'] == ['Unknown key.']
```

ramnes replied:
fixed, thanks

sloria commented on marshmallow/marshalling.py:
```diff
@@ -275,6 +283,22 @@ def deserialize(self, data, fields_dict, many=False, partial=False,
+                    set_value(ret, key, value)
+                elif unknown == RAISE:
+                    def _raise(_):
+                        raise ValidationError('Unkown key.')
```
Typo here: "Unknown". Also, we might want to make the error message `'Unknown field.'` to make it more language-agnostic.

ramnes replied:
fixed

sloria commented on marshmallow/schema.py:
```diff
@@ -517,6 +526,10 @@ def loads(self, json_data, many=None, partial=None, **kwargs):
             the value for `self.partial` is used. If its value is an iterable,
             only missing fields listed in that iterable will be ignored.
         :return: A dict of deserialized data
+        :param unknown: Whether to include, ignore or alert for unknown
```
Minor revision suggestion:

```
Whether to include, ignore, or raise an error for unknown...
```

ramnes replied:
done

sloria commented on marshmallow/schema.py:
```diff
@@ -486,7 +490,7 @@ def dumps(self, obj, many=None, update_fields=True, *args, **kwargs):
         return self.opts.render_module.dumps(serialized, *args, **kwargs)
 
-    def load(self, data, many=None, partial=None):
+    def load(self, data, many=None, partial=None, unknown=None):
```
Is it necessary/useful to add `unknown` as a parameter to `load`, or will the constructor parameter suffice?

ramnes replied:
I've done it just like `partial` (and `many`), which I think are doing it the right way. I'd say that the more flex you can get, the better it is.

sloria reviewed (changes requested):
Initial pass finished.

Thanks for taking this on!

sloria commented on marshmallow/marshalling.py:
```diff
@@ -275,6 +283,22 @@ def deserialize(self, data, fields_dict, many=False, partial=False,
+                        getter_func=_raise,
+                        data=value,
+                        field_name=key,
+                        field_obj=type(value)
```
`field_obj` is supposed to be the `Field` instance associated with the invalid key. In this case, there isn't a `field_obj`, so `call_and_store` should probably be modified to allow `field_obj` to be omitted.

ramnes commented:
Thanks a lot for your fast review, it answered most of the questions I had. I'll iterate later today, and the result should be much better than this.

lafrech reviewed (commented):
I never used that review feature before. I entered a few review comments last night but did not submit correctly. Here they are. Some of them overlap with @sloria's.

sloria commented:
> Question is, what do we want to store in the ValidationError fields and field_names? Currently, we only have field errors or schema errors. Error for unknown keys are not really schema errors, but they don't correspond to a field either. Passing the data key as field name makes sense but I don't know what to pass as field object.

My initial thought was to append the field name but not append the field object, but that might not be correct. Since we're saying the fields are "unknown", they probably shouldn't be stored. Also, `field_names` and `fields` should probably be kept the same length to make indexing into them easier.

> If we decide not to append to fields and field_names for unknown key error, then user can't rely on their length being zero to assert there was no error. (But who would do that anyway?)

I don't think that's a case we should handle. The user gets an exception when there are errors, so there's no need to check the length of `fields` or `field_names`.

ramnes commented:
Also, ignore or raise by default?

ramnes commented:
And in the end, I think that it would be great to have this option available in `Meta` too, but still overridable at instanciation and/or at load time. That would give a lot of flexibility, and thus things like this:

```python
from marshmallow import ALLOW, RAISE
from marshmallow import Schema as BaseSchema


# define a default class because in our application
# we want to raise an error by default
class Schema(BaseSchema):
    class Meta:
        unknown = RAISE


# define some schemas
class FooSchema(Schema):
    ...


def some_context(data, ...):
    schema = FooSchema()
    if ...:
        # oh wait, for this schema in that particular
        # context, we want to allow unknown keys
        schema.load(data, unknown=ALLOW)
```

ramnes committed:
Put the error storage logic in its own method

... so that we can store an error directly.

ramnes commented:
All requested changes are done.

ramnes commented:
> My initial thought was to append the field name but not append the field object, but that might not be correct. Since we're saying the fields are "unknown", they probably shouldn't be stored. Also, field_names and fields should probably be kept the same length to make indexing into them easier.

I'm just realizing that the naive method you're describing here at first is how I've done it. If the `field_names` aren't stored for unknown fields, how is the user supposed to know that there's a problem with the data?

lafrech commented:
In `'raise'` mode, an `Exception` is raised and the faulty keys appear in `'messages'`. In `'allow'` or `'ignore'` mode, there is no problem.

sloria commented on marshmallow/schema.py:
```diff
@@ -517,6 +526,10 @@ def loads(self, json_data, many=None, partial=None, **kwargs):
+        :param unknown: Whether to include, ignore, or raise an error for unknown
+            fields in the data. Use `ALLOW`, `IGNORE` or `RAISE`.
+            If `None`, the value for `self.unknown` is used.
+        :return: A tuple of the form (``data``, ``errors``)
```
This line should no longer be here.

ramnes replied:
Woops, another rebase error, sorry about this. It's fixed, thanks.

sloria commented:
Looking into the `fields` and `field_names` attributes further, I don't think they're actually necessary, so I've opened #840 to propose removing them. That said, that issue doesn't need to block merging this PR, and I think the approach you've taken (adding the unknown fields to `field_names`) is fine.

I'll look at the rest of the PR more closely when I get a chance.

ramnes commented:
I've just added a commit so that `unknown` is also available in `Schema.Meta`.

@sloria how quick do you see this in the tree? We really want to take advantage of this at @numberly, and we have a technological iteration sprint on our applications coming in a week or so. I'd love to see this feature to be merged meanwhile so that we could add it to our sprint checklist; do you think that's feasible?

sloria commented:
@ramnes Thanks for adding the `Meta` option. 👍 

I can't promise that we'll get this in by the time you need it (we--the maintainers--are volunteering free time, without a set schedule). That said, this is looking good so far; I could see it landing soon. I'll cut a release after this is merged so that you can start using it right away.

sloria commented:
There are a few housekeeping things that need to be done:

- [ ] Replace the existing example in the docs: https://marshmallow.readthedocs.io/en/latest/extending.html#validating-original-input-data
- [ ] Add a new blurb in the upgrading guide about this feature

@ramnes Feel free to do any of these if you have time.

sloria commented on tests/test_deserialization.py:
```diff
@@ -1315,6 +1315,77 @@ class MySchema(Schema):
+        assert 1 in err.messages
+        assert 'bar' in err.messages[1]
+        assert err.messages[1]['bar'] == ['Unknown field.']
+
```
Can you also add a test that `unknown` behaves correctly when `data_key` is set for a field?

ramnes replied:
Well seen, it indeed doesn't work properly. I'm digging on it.

ramnes replied:
Fixed. :+1:

ramnes commented:
Excellent, I'll do those changes, thank you for writing them down.

And by the way, you really don't need to justify yourself, I know how open source works; huge props to you for that voluntary work and for the quick answers!

ramnes commented:
Reading `upgrading.rst`, I feel that it's mostly the place where breaking changes are identified, and I don't feel that we're adding one here; unless we choose to raise by default, obviously. Should I modify `CHANGELOG.rst` first and foremost to explain the feature as a whole?

sloria commented:
Yeah, go ahead and start with the changelog. We can update the upgrading as necessary later.

ramnes commented:
There you go. I've added some documentation in the quickstart, moved and modifed things a bit in the `extending.rst` so that it's more logical given this modification, and added an entry to the changelog.

sloria commented:
Thanks for updating the docs.

Looks like we have a [unanimous vote](https://github.com/marshmallow-code/marshmallow/issues/524#issuecomment-397165731) for `RAISE` as the default. Can you change that here? Also, can you rename `ALLOW` -> `PASS` and `IGNORE` -> `DROP`, per https://github.com/marshmallow-code/marshmallow/issues/524#issuecomment-397413038 ?

ramnes commented:
Rename done. :+1: 

As for making `RAISE` the default, it's a bit trickier: it puts 21 tests in error… I feel it would overload this PR if I make all the changes required for this API change. Do you think we could make it in a second time, in another PR?

ramnes commented:
Also, does it still make sens to have the `pass_original` option if we raise by default? :thinking:

lafrech commented:
> I feel it would overload this PR if I make all the changes required for this API change. Do you think we could make it in a second time, in another PR?

Yes, absolutely (IMHO).

sloria commented:
Agreed. We can defer changing the default to `RAISE`.

As for `pass_original`: I think it's still needed. Users may need to do processing based on the original data, regardless of whether extra keys were passed.

sloria commented on tests/test_deserialization.py:
```diff
@@ -1387,6 +1387,23 @@ class Meta:
+
+        data = MySchema(unknown=RAISE).load({"Foo": 1})
+        assert data["foo"] == 1
+        assert "Foo" not in data
```
Minor: Can you also add an assertion that `foo` is considered an unknown field and `MySchema(unknown=RAISE).load({"foo": 1})` should therefore raise an error?

ramnes replied:
Done

nicktimko commented:
How about `INCLUDE` instead of `PASS`?  I'll stop bikeshedding...

trubesv commented on docs/quickstart.rst:
```diff
@@ -338,6 +338,45 @@ Or you can ignore missing fields entirely by setting ``partial=True``.
+
+By default, :meth:`load <Schema.load>` will drop any field that has not defined in the schema.
+
+This behavior can be modified with the ``unknown`` option, which accept three values:
```
accept**s***

ramnes replied:
Done

trubesv commented on docs/quickstart.rst:
```diff
@@ -338,6 +338,45 @@ Or you can ignore missing fields entirely by setting ``partial=True``.
+Handling Unknown Fields
++++++++++++++++++++++++
+
+By default, :meth:`load <Schema.load>` will drop any field that has not defined in the schema.
```
has not been*

ramnes replied:
Done

sloria commented:
> How about INCLUDE instead of PASS? I'll stop bikeshedding...

I was actually thinking the same thing: INCLUDE might be clearer than PASS. Not a big deal, though, and it's certainly not a blocker to getting this merged.

ramnes commented:
Want me to rename to `INCLUDE`, `EXCLUDE` and `RAISE`? I actually hesitated to propose that, because I didn't want to nitpick on the naming, but since we're here...

sloria commented:
Sure. @lafrech feel free to chime in, but I think INCLUDE, EXCLUDE and RAISE are the clearest names proposed thus far.

lafrech commented:
I like drop/pass but exclude/include makes total sense as well. Fine. let's do this.

![Let's do this.](http://gifimage.net/wp-content/uploads/2018/04/lets-do-this-gif-4.gif)

ramnes commented:
Rename done. I still have to fix @trubesv typos and to add the test you asked, @sloria. Do you see anything left after that?

sloria commented on CHANGELOG.rst:
```diff
@@ -4,10 +4,18 @@ Changelog
+Features:
+
+- The behavior to apply when encountering unknown fields while deserializing can be controlled with the ``unknown`` option (:issue:`524`).
+  It makes it possible to either "drop", "pass" or "raise".
```
This needs to be updated to "exclude", "include", or "raise".

ramnes replied:
Done

sloria commented on docs/quickstart.rst:
```diff
@@ -338,6 +338,45 @@ Or you can ignore missing fields entirely by setting ``partial=True``.
+
+This behavior can be modified with the ``unknown`` option, which accept three values:
+
+- `DROP <marshmallow.utils.DROP>`, which is the default
```
These need to be updated.

sloria replied:
Also, this should explain what the option does, for consistency.

```
- `EXCLUDE <marshmallow.utils.EXCLUDE>`, to exclude unknown fields from the returned data. This is the default behavior.
```

ramnes replied:
Done

sloria commented on docs/quickstart.rst:
```diff
@@ -338,6 +338,45 @@ Or you can ignore missing fields entirely by setting ``partial=True``.
+
+The ``unknown`` option value set in :meth:`load <Schema.load>` will always override the value applied at instantiation time, which itself will override the value defined in the *class Meta*.
+
+This precedence allows you to define your own favorite behavior and to override it in a different context that could require a different behavior.
```
Revision suggestion:

```
This order of precedence allows you to change the behavior of a schema for different contexts.
```

ramnes replied:
Done

sloria reviewed (commented):
Just a few more revision suggestions to the docs. Once the remaining nits are addressed, this should be good to merge!

sloria commented on docs/quickstart.rst:
```diff
@@ -338,6 +338,45 @@ Or you can ignore missing fields entirely by setting ``partial=True``.
+
+    class UserSchema(Schema):
+        class Meta:
+            unknown = PASS
```
Update these to `INCLUDE`.

ramnes replied:
Done

sloria commented on docs/extending.rst:
```diff
@@ -37,6 +37,36 @@ By default, pre- and post-processing methods receive one object/datum at a time,
+Using Original Input Data
++++++++++++++++++++++++++
+
+By default, unspecified field names are ignored by the validator. If for some reason you want to use the original, raw input, you can add ``pass_original=True`` to your call to `post_load <marshmallow.decorators.post_load>`.
```
Minor revison suggestion:

```
If you want to use the original, unprocessed input, 
you can add pass_original=True` to your call to post_load.
```

ramnes replied:
Done

ramnes commented:
Woops, looks like I forgot to update the docs when renaming from `PASS` / `DROP` to `INCLUDE` / `EXCLUDE`. You could have made only one comment for the whole documentation. :wink: 

I've done everything here. I've also simplified the history back to its original state because rebases where becoming more and more painful with all the extra context.

sloria commented:
Looks great! I think this is good to merge. It would be nice to have a way to customize the error message, but that can happen in a follow-up PR. Let's get this merged.

lafrech commented:
A bit late to the party but if we want to enforce the use of the `EXCLUDE`, `INCLUDE` and `RAISE` singletons defined in `utils.py`, we could use `is` and `is not` rather than `==` and `!=` when testing the value of `unknown`.

This implementation allows the user to pass strings directly, although it is not explicitly supported.

sloria commented:
@lafrech I don't see a benefit to enforcing usage of the constants. It should suffice to document their usage. Also, Sphinx will render the default as a string.

<img width="600" alt="api reference marshmallow 3 0 0b12 dev0 documentation 2018-07-01 13-39-38" src="https://user-images.githubusercontent.com/2379650/42137086-50858af4-7d34-11e8-92fb-f60c3387be54.png">

lafrech commented:
That could arguably be a Sphinx issue. Anyway, I don't mind if we leave it as it is.

ramnes commented:
@sloria may I ask you to tag the current HEAD so that we can profit from this merge?

deckar01 commented:
There are some followup issues that have not been resolved. (#851, #852)

`pip` can install packages directly from github. https://pip.pypa.io/en/stable/reference/pip_install/#git

I believe the URL would be `git+https://github.com/marshmallow-code/marshmallow#egg=marshmallow`.

ramnes commented:
> pip can install packages directly from github. 

Yeah @deckar01, I know that. We just want to go back on the PyPI version and stop dealing with pip eggs, which we use for very precise things and imply more work for us.

> There are some followup issues that have not been resolved. (#851, #852)

@sloria seemed okay to have a version tagged as soon as this was merged (https://github.com/marshmallow-code/marshmallow/pull/838#issuecomment-396917731, https://github.com/marshmallow-code/marshmallow/pull/838#issuecomment-397842618), hence my ping.

sloria commented:
There were a few bugs that were surfaced with this feature, so I wanted give this a little more time to "bake". But those bugs have been addressed. The followup issues above can be deferred to a near-future release. So yeah, let's ship it!

ramnes commented:
Awesome, thank you! :tada:
